CPMemberAddRemoveTest.when_cpMembersShutdownConcurrently_then_theyCompleteTheirShutdown

http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/547/testReport/junit/com.hazelcast.cp.internal/CPMemberAddRemoveTest/when_cpMembersShutdownConcurrently_then_theyCompleteTheirShutdown/

```
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cp.internal.CPMemberAddRemoveTest$25.run(CPMemberAddRemoveTest.java:952)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1302)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1328)
	at com.hazelcast.cp.internal.CPMemberAddRemoveTest.when_cpMembersShutdownConcurrently_then_theyCompleteTheirShutdown(CPMemberAddRemoveTest.java:948)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

_master_ (commit 358724cfa52179609f38bb2c245f7bc310cf4ab4)

http://jenkins.hazelcast.com/job/Hazelcast-3.x-OracleJDK11/9/testReport/com.hazelcast.cp.internal/CPMemberAddRemoveTest/when_cpMembersShutdownConcurrently_then_theyCompleteTheirShutdown/

**Stacktrace**
```
java.lang.AssertionError: Future has not completed
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport$7.run(HazelcastTestSupport.java:1023)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1307)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1333)
	at com.hazelcast.test.HazelcastTestSupport.assertCompletesEventually(HazelcastTestSupport.java:1020)
	at com.hazelcast.cp.internal.CPMemberAddRemoveTest.when_cpMembersShutdownConcurrently_then_theyCompleteTheirShutdown(CPMemberAddRemoveTest.java:973)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

--- a/hazelcast/src/main/java/com/hazelcast/cp/CPSubsystemManagementService.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/CPSubsystemManagementService.java
@@ -17,6 +17,7 @@
 package com.hazelcast.cp;
 
 import com.hazelcast.config.cp.CPSubsystemConfig;
+import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.ICompletableFuture;
 import com.hazelcast.cp.exception.CPGroupDestroyedException;
 
@@ -104,6 +105,36 @@ import java.util.Collection;
  * additional 4 regular Hazelcast members should exist in the cluster.
  * New Hazelcast members can be started to satisfy
  * {@link CPSubsystemConfig#getCPMemberCount()}.
+ * <p>
+ * <strong>There is a subtle point about graceful shutdown of CP members.
+ * If there are N CP members in the cluster, {@link HazelcastInstance#shutdown()}
+ * can be called on N-2 CP members concurrently. Once these N-2 CP members
+ * complete their shutdown, the remaining 2 CP members must be shut down
+ * serially.
+ * <p>
+ * Even though the shutdown API is called concurrently on multiple members,
+ * the Metadata CP group handles shutdown requests serially. Therefore,
+ * it would be simpler to shut down CP members one by one, by calling
+ * {@link HazelcastInstance#shutdown()} on the next CP member once the current
+ * CP member completes its shutdown.
+ * <p>
+ * The reason behind this limitation is, each shutdown request internally
+ * requires a Raft commit to the Metadata CP group. A CP member proceeds to
+ * shutdown after it receives a response of its commit to the Metadata CP
+ * group. To be able to perform a Raft commit, the Metadata CP group must have
+ * its majority available. When there are only 2 CP members left after graceful
+ * shutdowns, the majority of the Metadata CP group becomes 2. If the last 2 CP
+ * members shut down concurrently, one of them is likely to perform its Raft
+ * commit faster than the other one and leave the cluster before the other CP
+ * member completes its Raft commit. In this case, the last CP member waits for
+ * a response of its commit attempt on the Metadata group, and times out
+ * eventually. This situation causes an unnecessary delay on shutdown process
+ * of the last CP member. On the other hand, when the last 2 CP members shut
+ * down serially, the N-1th member receives response of its commit after its
+ * shutdown request is committed also on the last CP member. Then, the last CP
+ * member checks its local data to notice that it is the last CP member alive,
+ * and proceeds its shutdown without attempting a Raft commit on the Metadata
+ * CP group.</strong>
  *
  * @see CPMember
  * @see CPSubsystemConfig

--- a/hazelcast/src/test/java/com/hazelcast/cp/internal/CPMemberAddRemoveTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cp/internal/CPMemberAddRemoveTest.java
@@ -929,15 +929,18 @@ public class CPMemberAddRemoveTest extends HazelcastRaftTestSupport {
 
     @Test
     public void when_cpMembersShutdownConcurrently_then_theyCompleteTheirShutdown() throws ExecutionException, InterruptedException {
-        // When there are N CP members, we can perform concurrent shutdown in 2 steps.
+        // When there are N CP members, we can perform partially-concurrent shutdown in 2 steps:
         // In the first step, we shut down N - 2 members concurrently.
-        // Once those members are done, we shutdown the last 2 CP members concurrently as well.
+        // Once those members are done, we shutdown the last 2 CP members serially.
+        // The last 2 CP members must be shutdown serially because if both of them shutdown at the same time,
+        // one of them can commit its leave to the Metadata group and terminate before the other one performs its commit.
+        // In this case, the last member hangs since there is no available majority of the Metadata group anymore.
 
         final HazelcastInstance[] instances = newInstances(7, 5, 0);
 
-        final int batchSize = 5;
-        Future[] futures = new Future[batchSize];
-        for (int i = 0; i < batchSize; i++) {
+        final int concurrent = 5;
+        Future[] futures = new Future[concurrent];
+        for (int i = 0; i < concurrent; i++) {
             final int ix = i;
             futures[i] = spawn(new Runnable() {
                 @Override
@@ -952,21 +955,8 @@ public class CPMemberAddRemoveTest extends HazelcastRaftTestSupport {
             f.get();
         }
 
-        int remaining = instances.length - batchSize;
-        futures = new Future[remaining];
-        for (int i = 0; i < remaining; i++) {
-            final int ix = i;
-            futures[i] = spawn(new Runnable() {
-                @Override
-                public void run() {
-                    instances[batchSize + ix].shutdown();
-                }
-            });
-        }
-
-        for (Future f : futures) {
-            assertCompletesEventually(f);
-            f.get();
+        for (int i = 0, remaining = (instances.length - concurrent); i < remaining; i++) {
+            instances[concurrent + i].shutdown();
         }
     }
 

